### PR TITLE
Add kfctl E2E test on GCP with IAP

### PR DESCRIPTION
**Pull request by jlewi**
Related to #2610 E2E test for go binary

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/kubeflow/2814)

### CODE REVIEW

**Comment by jlewi** [comment]
#2795  should be submitted first.

**Comment by jlewi** [comment]
Most recent failure was problem creating sandbox: https://github.com/kubeflow/testing/issues/342

**Comment by jlewi** [comment]
Test passed!

**Comment by jlewi** [comment]
@gabrielwen @kkasravi This should be ready to review.

File: testing/kfctl/kfctl_go_test.py [hunk 1]
```diff
@@ -71,10 +80,9 @@ def test_build_kfctl_go(app_path, project):
   # username and password are passed as env vars and won't appear in the logs
   run_with_retries([kfctl_path, "init", app_path, "-V", "--platform=gcp",
                          "--version=" + version,
```
Comment (kkasravi) [inline testing/kfctl/kfctl_go_test.py]: should the PULL_NUMBER be used to fetch that PR from github
eg `kfctl init ... --version pull/os.getenv("PULL_NUMBER")

**Comment by jlewi** [comment]
Latest test failure is:

```
INFO Error: couldn't apply KfApp: coordinator Apply failed for gcp: gcp apply could not update deployment manager Error could not update cluster-kubeflow.yaml: Update deployment error: googleapi: Error 409: Resource 'projects/kubeflow-ci-deployment/global/deployments/kfctl-e58c' has an ongoing conflicting operation: 'projects/kubeflow-ci-deployment/global/operations/operation-1553727977511-5851b7c2d5878-c8dabac2-0ec03ba6'., conflict
```

But it looks like the pod actually got evicted because of the deadline.

```
Pod was active on the node longer than the specified deadline
```

File: testing/kfctl/kfctl_go_test.py [hunk 2]
```diff
@@ -61,8 +61,17 @@ def test_build_kfctl_go(app_path, project):
...
-  os.environ["KUBEFLOW_USERNAME"] = "kf-test-user"
-  os.environ["KUBEFLOW_PASSWORD"] = str(uuid.uuid4().hex)
+  init_args = []
+  if use_basic_auth:
+    os.environ["KUBEFLOW_USERNAME"] = "kf-test-user"
+    os.environ["KUBEFLOW_PASSWORD"] = str(uuid.uuid4().hex)
+    init_args = ["--use_basic_auth"]
+  else:
+    # Owned by project kubeflow-ci-deployment.
+    os.environ["CLIENT_ID"] = "CJ4qVPLTi0j0GJMkONj7Quwt"
```
[the remark this answers is not in the document]
Comment (jlewi) [reply testing/kfctl/kfctl_go_test.py]: Yes. CLIENT id's can only be used with specific URLs which we own and configure.
CLIENT_ID & SECRET aren't actually secret. e.g. if you use them in a webapp in client side JS they are visible.

File: testing/kfctl/kfctl_go_test.py (hunk 1 above)
Comment (jlewi) [reply testing/kfctl/kfctl_go_test.py]: Isn't that what's happening? 
Although once https://github.com/kubeflow/kubeflow/issues/2831 we should just pull the configs from the local checkout of the source.

**Comment by jlewi** [comment]
Woo Hoo! Tests are passing again. I had to increase the timeout.

**Comment by gabrielwen** [comment]
> Woo Hoo! Tests are passing again. I had to increase the timeout.

yea... timeout is becoming an issue on my end as well...

**Comment by jlewi** [comment]
@gabrielwen could you LGTM please?

**Comment by jlewi** [comment]
Recent test should be a flake.

```
util.py                     71 INFO     Error: couldn't delete KfApp: coordinator Delete failed for gcp: Error when cleaning IAM policy: googleapi: Error
409: There were concurrent policy changes. Please retry the whole read-modify-write with exponential backoff., aborted
```

/test all

**Pull request merged**